datafusion/physical-plan/src/joins/nested_loop_join.rs

```diff
@@ -100,6 +101,144 @@ impl JoinLeftData {
+/// Output batches are either combined or split to ensure that the
+/// batches produced are as close to the target size as possible
+/// without being any bigger.
+struct OutputBuffer {
+    /// Whether the buffer has output any record batches,
+    /// used to ensure that at least one record batch is
+    /// produced even if no joins happen.
+    have_output: bool,
+    /// Output batch
+    batches: VecDeque<RecordBatch>,
+    /// Number of rows in the buffer
+    total_rows: usize,
+    /// Batch that is ready to be output.
+    next_batch: Option<(RecordBatch, usize)>,
+    /// Schema of the output batches
+    schema: SchemaRef,
+    /// Target output batch size
+    target_batch_size: usize,
+    /// Memory reservation for tracking batch memory use
+    memory_reservation: MemoryReservation,
+}
+
+impl OutputBuffer {
+    fn new(
+        schema: SchemaRef,
+        target_batch_size: usize,
+        memory_reservation: MemoryReservation,
+    ) -> Self {
+        Self {
+            have_output: false,
+            batches: VecDeque::new(),
+            total_rows: 0,
+            next_batch: None,
+            schema,
+            target_batch_size,
+            memory_reservation,
+        }
+    }
+
+    /// Pash a new batch into the output buffer.
+    fn push(&mut self, batch: RecordBatch) -> Result<()> {
+        assert!(self.next_batch.is_none());
+        assert!(self.total_rows < self.target_batch_size);
+
+        self.memory_reservation
+            .try_grow(batch.get_array_memory_size())?;
+        if self.total_rows + batch.num_rows() < self.target_batch_size {
+            // Not enough rows to fill a batch.
+            self.total_rows += batch.num_rows();
+            self.batches.push_back(batch);
+            return Ok(());
+        }
+        let mut batches = std::mem::take(&mut self.batches);
+        if self.total_rows + batch.num_rows() == self.target_batch_size {
+            self.total_rows = 0;
+            batches.push_back(batch);
+        } else {
+            self.total_rows = batch.num_rows();
+            self.batches.push_back(batch);
+        }
+        let mut total_array_memory_size = 0;
+        let batch = concat_batches(
+            &self.schema,
+            batches.iter().inspect(|batch| {
+                total_array_memory_size += batch.get_array_memory_size()
+            }),
+        )?;
+        self.next_batch = Some((batch, total_array_memory_size));
+        Ok(())
+    }
+
+    /// Return a batch if there are enough rows in the buffer.
+    fn next(&mut self) -> Option<RecordBatch> {
+        if let Some((batch, total_array_memory_size)) = self.next_batch.take() {
+            self.have_output = true;
+            self.memory_reservation.shrink(total_array_memory_size);
+            return Some(batch);
+        }
+        if self.total_rows < self.target_batch_size {
+            return None;
+        }
+        if self.total_rows == self.target_batch_size {
```

Comment:
I wonder if we are guaranteed that the rows will always be exactly the target batch size? 

Maybe it it would be safer to use something like

```suggestion
        if self.total_rows >= self.target_batch_size {
```